### PR TITLE
ipopt 3.12.7

### PR DESCRIPTION
**Pull request by goxberry**
The commits for #4489 and #4482 are mingled into this commit also; need to figure out a way to separate them. Anyway, the last three commits constitute the ipopt changes, and these changes depend on merging #4482 and #4489.

### CODE REVIEW

**Comment by davydden** [comment]
> depend on merging #4482 and #4489.

once those are merged, please rebase on `develop`

**Comment by adamjstewart** [comment]
> The commits for #4489 and #4482 are mingled into this commit also; need to figure out a way to separate them. 

I think that's what [cherry-picking](http://spack.readthedocs.io/en/latest/contribution_guide.html#cherry-picking) is for. @davydden can help you if you run into trouble (he wrote those docs). Personally I create a new branch for every package I add, but I understand that this can be difficult when you depend on other branches to write your package.

**Comment by davydden** [comment]
@goxberry you can also do `git rebase -i develop` and then drop first three commits during rebasing. Then double check that everything is ok and do `git push --force` to your github remote.

File: var/spack/repos/builtin/packages/ipopt/package.py [hunk 1]
```diff
@@ -68,6 +78,20 @@ def install(self, spec, prefix):
             "--with-lapack-lib=%s" % lapack_lib
         ]
 
+        if 'coinhsl' in spec:
+            coinhsl_dir = spec['coinhsl'].prefix
+            coinhsl_lib = '-L%s -lcoinhsl' % coinhsl_dir.lib
```
Comment (davydden) [inline var/spack/repos/builtin/packages/ipopt/package.py]: you can use direction `spec['coinhsl'].libs.ld_flags`

File: var/spack/repos/builtin/packages/ipopt/package.py [hunk 2]
```diff
@@ -68,6 +78,20 @@ def install(self, spec, prefix):
...
+            configure_args.extend([
+                '--with-hsl-lib=%s' % coinhsl_lib,
+                '--with-hsl-incdir=%s' % coinhsl_dir.include])
+
+        if 'metis' in spec:
+            metis_dir = spec['metis'].prefix
+            metis_lib = '-L%s -lmetis' % metis_dir.lib
```
Comment (davydden) [inline var/spack/repos/builtin/packages/ipopt/package.py]: same as above.

File: var/spack/repos/builtin/packages/ipopt/package.py (hunk 1 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/ipopt/package.py]: the default implementation of `libs` look for `l<packagename>` library, which is your case.

**Comment by goxberry** [comment]
@davydden Should be all set. Thanks for the advice! Those are all commands with which I'm familiar. I wouldn't cherry-pick these commits either. The main issue during development is that I have to rebase onto at least the `coinhsl` branch I wrote in order to test an `ipopt` install. If you're not worried about testing this PR until after the `coinhsl` (#4482) and `metis` (#4489) PRs are merged, it's easy enough for me to `git-rebase -i` off the commits from those PRs after locally testing.

File: var/spack/repos/builtin/packages/ipopt/package.py [hunk 3]
```diff
@@ -31,16 +31,26 @@ class Ipopt(Package):
...
+            description="Build with Coin Harwell Subroutine Libraries")
+    variant('metis', default=False,
+            description="Build with METIS partitioning support")
+
     depends_on("blas")
     depends_on("lapack")
     depends_on("pkg-config", type='build')
     depends_on("mumps+double~mpi")
+    depends_on('coinhsl', when='+coinhsl')
+    depends_on('metis@4.0.3', when='+metis')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/ipopt/package.py]: Is this really the only version of `metis` it works with?

Comment (goxberry) [inline var/spack/repos/builtin/packages/ipopt/package.py]: It only works with METIS 4; see https://www.coin-or.org/Ipopt/documentation/node13.html#SECTION00043700000000000000. I guess, theoretically, the spec could be `depends_on('metis@4')`, and I could try debugging that again, but I think spack returned an error with that particular spec, and `metis@4.0.3` is the only version of METIS 4 in spack.

It's possible to add METIS 4.0.1 to spack, and then test a more general spec, but is doing so really worth it?

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/ipopt/package.py]: I know this is kind of confusing, and I would like to change this behavior as well, but `metis@4` means _exactly_ version 4. What you really want is `metis@4.0:4.999`.

**Comment by goxberry** [comment]
@adamjstewart Made the version change. Thanks for the clarification!

**Pull request merged**